cf-agent/cf-agent.c

```diff
@@ -238,11 +238,21 @@ int main(int argc, char *argv[])
     {
         Log(LOG_LEVEL_ERR, "CFEngine was not able to get confirmation of promises from cf-promises, so going to failsafe");
         EvalContextClassPutHard(ctx, "failsafe_fallback", "attribute_name=Errors,source=agent");
-        GenericAgentConfigSetInputFile(config, GetInputDir(), "failsafe.cf");
-        policy = LoadPolicy(ctx, config);
+        
+        if (CheckAndGenerateFailsafe(GetInputDir(), "failsafe.cf"))
+        {
+            GenericAgentConfigSetInputFile(config, GetInputDir(), "failsafe.cf");
+            policy = LoadPolicy(ctx, config);
+        }
     }
     assert(policy);
-
+    
+    if (!policy)
```

Comment:
> So assert() should be redundant as well

Yes, all asserts are redundant. It should **never** trigger. However we put them there to ensure we don't forget our own coding assumptions.

This is the way I'm deciding for using asserts:
* Does the check look redundant? Use `assert()`, it won't cost in runtime for non-debug builds anyway.
* Does the check look necessary? I.e. is there one in a million it will trigger? Then do proper error-handling.